Catch NoMethodErrors in FEDEX find_and_parse_rate method

heinzf opened the pull request:
See https://github.com/Shopify/active_shipping/issues/307

The ```parse_rate_response``` method would crash and return a NoMethodError whenever the service type was missing in the XML response from FEDEX.

This seems to be related to the new version of the upstream API. As a workaround before we implement this new version, this PR proposes to catch all NoMethodErrors in ```the parse_rate_response``` loop through the rates, and return as much as possible.

Because of this, the rates list might contain nil values, hence it will be compacted and also kept as successful, as not to raise an exception afterward, while logging the faulty XML response.

christianblais commented on lib/active_shipping/carriers/fedex.rb:
```diff
@@ -475,7 +475,8 @@ def parse_rate_response(origin, destination, packages, response, options)
         rate_estimates = xml.root.css('> RateReplyDetails').map do |rated_shipment|
-          service_code = rated_shipment.at('ServiceType').text
+          service_code = rated_shipment.at('ServiceType').try(:text)
+          raise ActiveShipping::ResponseError, "The response did not contain service type" unless service_code.present?
```
Since it is a change in their API, I'm afraid this will just continue to the next attribute that isn't gonna be define neither. Do we have a copy of the response we're receiving? Could we run a test against it?

heinzf replied:
Ah I see what you mean now. I'll try to find it to see if there's something else broken you're right

mdking replied:
As mentioned in slack I would offer up the suggestion of rather than directly raising the `ActiveShipping::ResponseError`, create a `RateResponse` with `success = false` similar to how it is performed for no rates here https://github.com/heinzf/active_shipping/blob/master/lib/active_shipping/carriers/fedex.rb#L500

This way the `ActiveShipping::ResponseError` will have the failed `RateResponse` with the xml and request. Would just want to verify that those fields are being logged somewhere.

heinzf commented:
@christianblais @mdking 
I changed things a bit following your comments.
I will try to "catch" missing service_type fields and log the xml. The rates list will then be compacted and kept as successful, as not to raise an exception afterward.

The idea is to log the bad behaving XML in a warning, but still give as much information as possible, as a best effort.

Tell me if the idea and the flow seem legit.

mdking commented on lib/active_shipping/carriers/fedex.rb:
```diff
@@ -495,15 +500,23 @@ def parse_rate_response(origin, destination, packages, response, options)
+        if missing_xml_field && logger
+          logger.warn(xml)
+        end
+
```
I would `compact` once here on the `rate_estimates` it doesn't have to be repeated.

mdking commented on lib/active_shipping/carriers/fedex.rb:
```diff
@@ -495,15 +500,23 @@ def parse_rate_response(origin, destination, packages, response, options)
+          logger.warn(xml)
+        end
+
+        if rate_estimates.compact.empty? && missing_xml_field
```
This can be rewritten:

```
if rates_estmates.empty?
  success = false
  
  if miss_xml_field
    ...
  else 
    ...
  end
end

mdking commented on test/unit/carriers/fedex_test.rb:
```diff
@@ -205,6 +205,32 @@ def test_parsing_response_with_no_rate_reply
+    end
+    message = "The response from the carrier contained errors and could not be treated"
+    assert_equal exception.message, message
+  end
```
Can you add the log assert as part of this test?

heinzf replied:
I've been trying for a while with no success yet :(

christianblais commented on lib/active_shipping/carriers/fedex.rb:
```diff
@@ -472,10 +472,15 @@ def parse_rate_response(origin, destination, packages, response, options)
+          unless service_code.present?
+            missing_xml_field = true
+            next
+          end
```
Since this is a change in the API, I see that other errors could happen later on in the process, such as `rated_shipment.at('RatedShipmentDetails/ShipmentRateDetail/TotalNetCharge/Currency').text`. Since we probably don't want to wrap all the calls with a `if` like the above, should we simply rescue `NoMethodError`, in which we could log?

heinzf replied:
@christianblais Indeed.
I've added a rescue/begin and kept the same "return what we can" logic

mdking commented on test/unit/carriers/fedex_test.rb:
```diff
@@ -205,6 +205,31 @@ def test_parsing_response_with_no_rate_reply
+
+    @carrier.expects(:commit).returns(mock_response)
+    @carrier.logger = Logger.new(StringIO.new)
+    @carrier.logger.expects(:warn)
```
Thanks for adding that, if you would like to be more explicit you can add:
`@carrier.logger.expects(:warn).with(mock_response).once`

christianblais commented on lib/active_shipping/carriers/fedex.rb:
```diff
@@ -474,31 +474,46 @@ def parse_rate_response(origin, destination, packages, response, options)
+                 :delivery_range => delivery_range)
+          rescue NoMethodError
+            missing_xml_field = true
+            next
```
Is this required?

heinzf replied:
The ```next``` statement? Without it it the iteration returns the assignment to missing_xml_field, and the ```compact``` method afterward cannot do its job.
That being said, ```next``` is overkill as we are already at the end of the block and we'll start the next iteration anyway, a simple nil will do the trick.

christianblais commented on lib/active_shipping/carriers/fedex.rb:
```diff
@@ -474,31 +474,46 @@ def parse_rate_response(origin, destination, packages, response, options)
         end
 
+        rate_estimates = rate_estimates.compact
+        logger.warn(xml) if logger && missing_xml_field
```
I believe simply outputing `xml` might be confusing. Should we at least indicate there has been an error of some sort?

heinzf commented:
@christianblais @mdking 
I've added a log message and replace the if by a larger begin/rescue

christianblais commented:
Can you please provide a good PR description and meaningful commits please?

mdking commented:
Code LGTM. 

On our end when exception occurs for the `NoMethodError` it will be rescued and the response is logged. To log the warning our code will have to be modified to initialize the logger object when the carrier adapter is created.

kmcphillips commented on lib/active_shipping/carriers/fedex.rb:
```diff
@@ -474,31 +474,46 @@ def parse_rate_response(origin, destination, packages, response, options)
+          end
         end
 
+        rate_estimates = rate_estimates.compact
```
What is this for? This is a behaviour change.

heinzf replied:
I've updated the PR description to explain, it was in a comment bellow it :)

The method would loop through the XML elements for the rates with a ```array.map```.
If one of the iterations went through the ```begin/rescue```, the element would take the last value returned by the ```rescue```, which seems to be ```true```.
I've changed it to return ```nil``` so that we can compact the list and only keep the rates that were correctly parsed.

kmcphillips replied:
Thanks for clarifying.

kmcphillips commented:
One comment. But mostly a big improvement. Thank you!

heinzf commented:
Is it :rocket: ready?

heinzf commented:
@mdking @kmcphillips @christianblais 
Sorry, still not used to the correct github ping policy :(

christianblais commented on test/unit/carriers/fedex_test.rb:
```diff
@@ -205,6 +205,31 @@ def test_parsing_response_with_no_rate_reply
+
+    @carrier.expects(:commit).returns(mock_response)
+    @carrier.logger = Logger.new(StringIO.new)
+    @carrier.logger.expects(:warn).once
```
Should expect with what. `expects(:warn).once.with(BLA BLA)`

christianblais commented on test/unit/carriers/fedex_test.rb:
```diff
@@ -205,6 +205,31 @@ def test_parsing_response_with_no_rate_reply
+  end
+
+  def test_parsing_response_with_no_system_code
+    mock_response = xml_fixture('fedex/freight_rate_response').gsub('<v13:ServiceType>FEDEX_FREIGHT_ECONOMY</v13:ServiceType>','').gsub('<v13:ServiceType>FEDEX_FREIGHT_PRIORITY</v13:ServiceType>','')
```
Why twice?

christianblais commented on test/unit/carriers/fedex_test.rb:
```diff
@@ -205,6 +205,31 @@ def test_parsing_response_with_no_rate_reply
+    @carrier.logger.expects(:warn).once
+    exception = assert_raises(ActiveShipping::ResponseError) do
+      @carrier.find_rates( location_fixtures[:ottawa], location_fixtures[:beverly_hills], package_fixtures.values_at(:book, :wii), :test => true)
+
```
:scissors:

heinzf replied:
Damn, thanks.

heinzf replied:
Ah no sorry, I thought I had copy pasted one of the tests somewhere. There's one test per behaviour, whether there's at least one rate found, or none. (Returns something vs returns an error)

christianblais replied:
I was talking about the empty line.

christianblais commented on test/unit/carriers/fedex_test.rb:
```diff
@@ -205,6 +205,31 @@ def test_parsing_response_with_no_rate_reply
+
+    @carrier.expects(:commit).returns(mock_response)
+    @carrier.logger = Logger.new(StringIO.new)
+    @carrier.logger.expects(:warn).once
```
Let's be explicit about the `with`.

christianblais commented on test/unit/carriers/fedex_test.rb:
```diff
@@ -205,6 +205,31 @@ def test_parsing_response_with_no_rate_reply
+    assert_equal rates.rates[0].service_code, 'FEDEX_FREIGHT_PRIORITY'
+  end
+
+  def test_parsing_response_with_no_system_code
```
What does `system_code` mean?

heinzf commented:
@christianblais I've changed the title of the text to better explain that the system code belong to the shipping rate, made the empty lines cleaner, and added explicit validation for the content of the XML.

christianblais commented:
:+1:

kmcphillips commented:
Thanks @heinzf.

kmcphillips commented:
Released in 1.6.4
